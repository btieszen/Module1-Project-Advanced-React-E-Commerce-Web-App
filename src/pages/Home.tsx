import {useEffect} from 'react';
import { Product } from '../types/types';
import ProductCard from './ProductCard';
import './Home.css';
import { useProductContext } from '../context/ProductContext';
import {useQuery } from '@tanstack/react-query';
import { fetchProduct, fetchCategory } from '../api/api';

const Home = () => {
    const { state: { products, selectedCategory }, dispatch } = useProductContext();

  const {
     data: productsData,
 
     } =useQuery({
        queryKey:['products'],
        queryFn:fetchProduct,
    });

    const {data:categories} = useQuery({
        queryKey:['categories'],
        queryFn:fetchCategory,
    });

    useEffect(() => {
        if(productsData){
            dispatch({type:'SET_PRODUCTS', payload: productsData.data});
        }
    },[dispatch,productsData]);

    const getfilteredProducts = () => {
        if(selectedCategory) {
            return products?.filter(
                (product:Product) => product.category === selectedCategory
            );
        }
        return products;
    };

    const filteredProducts = getfilteredProducts();

    return (
        <div>
            <select
            onChange={(e)=> 
                dispatch({type:'SET_SELECTED_CATEGORY',payload:e.target.value})
            }
            >
                <option value=''>All Categories</option>
                {categories?.data.map((category) => (
                    <option value={category} key={category}>
                        {category}
                        </option>
                ))}
            </select>
        <div className='container'>
            {filteredProducts.map((product: Product) => (
            <ProductCard product={product} key ={product.id}/>
            ))}
        </div>
        </div>
    );
};
export default Home;