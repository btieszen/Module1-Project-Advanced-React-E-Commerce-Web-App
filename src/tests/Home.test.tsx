import { render } from '@testing-library/react';
import Home from '../pages/Home';
import { QueryClient,QueryClientProvider } from '@tanstack/react-query';
const mockNavigate = jest.fn()
const queryClient = new QueryClient()
// Mocking `useNavigate` from react-router-dom
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'), // Keeps other imports intact
    useNavigate: ()=>mockNavigate, // Mock `useNavigate`
  }));
test('matches snapshot',()=>{

    const {asFragment} = render(<QueryClientProvider client={queryClient}><Home /></QueryClientProvider>);
    
    expect(asFragment()).toMatchSnapshot();
});