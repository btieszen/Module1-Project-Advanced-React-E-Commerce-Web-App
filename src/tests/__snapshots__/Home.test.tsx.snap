// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <select>
      <option
        value=""
      >
        All Categories
      </option>
    </select>
    <div
      class="container"
    />
  </div>
</DocumentFragment>
`;
